stream_alert/rule_processor/rules_engine.py

```diff
@@ -470,22 +470,31 @@ def check_alerts_duplication(record, rule, alerts):
         Returns:
             bool: Return True if both record and rule name exist in alerts list.
         """
+        # Don't check normalization and threat intel keys
+        ignore_keys = {StreamThreatIntel.IOC_KEY, NORMALIZATION_KEY}
+        record_keys = set(record.keys()) - ignore_keys
         for exist_alert in alerts:
             if rule.rule_name == exist_alert['rule_name']:
-                record_copy = record.copy()
-                exist_alert_record_copy = exist_alert['record'].copy()
-                # Early return when two records are dumplicated.
-                if record_copy == exist_alert_record_copy:
-                    return True
+                alert_keys = set(exist_alert['record'].keys()) - ignore_keys
+
+                if alert_keys != record_keys:
+                    # If both keys set are different, skip value comparison and
+                    # continue
+                    continue
 
-                # There is chance that there are different values in 'streamalert:normalization'
-                # or 'streamalert:ioc' fields. So do comparision again after removing
-                # two keys.
-                record_copy.pop(StreamThreatIntel.IOC_KEY, None)
-                record_copy.pop(NORMALIZATION_KEY, None)
-                exist_alert_record_copy.pop(StreamThreatIntel.IOC_KEY, None)
-                exist_alert_record_copy.pop(NORMALIZATION_KEY, None)
-                if record_copy == exist_alert_record_copy:
+                found_duplication = False
```

Comment:
suggestion: could we simplify this and have this bool default to `True` then `break` like so:

```
found_duplication = True
for key in record_keys:
    if record[key] != exist_alert['record'][key]:
        found_duplication = False
        break
```

Reply:
I actually think we could improve this by inverting our logic.. so instead of checking if it IS a duplicate, let's check if it's _not_ a duplicate

This gives us the ability to exit early the moment we find a key/value that is NOT a match.. instead of having to look at ALL key/values to see that they all match before returning true. let's talk more about this before this merges :)

Reply:
I just did a quick test and inverse logic gets us a ~2x improvement

Reply:
I thought this at first too, but we already terminate early when a duplicate is found. We can't return early if a duplicate is *not* found, because there could still be a duplicate later on.

Since this is difficult to understand, @chunyong-lin I recommend moving the inner-loop logic into its own method which returns `True` if one alert matches the given record. Then the code will be more readable, and you can write lots of unit tests for that inner method to make sure its working as expected

Reply:
yep sorry @austinbyers and @chunyong-lin after discussion this morning I take all fo that back :)